src/compiler/checker.ts

```diff
@@ -8520,6 +8520,16 @@ namespace ts {
         function getContextuallyTypedParameterType(parameter: ParameterDeclaration): Type {
             const func = parameter.parent;
             if (isContextSensitiveFunctionOrObjectLiteralMethod(func)) {
+                const iife = getImmediatelyInvokedFunctionExpression(func);
+                if (iife) {
+                    const indexOfParameter = indexOf(func.parameters, parameter);
+                    if (iife.arguments && indexOfParameter < iife.arguments.length) {
+                        if (parameter.dotDotDotToken) {
+                            return createArrayType(getUnionType(map(iife.arguments.slice(indexOfParameter), getTypeOfExpression)));

```

Comment:
Switched to a for loop. As we discussed in person, it would be nice to have a map over a range for general use.